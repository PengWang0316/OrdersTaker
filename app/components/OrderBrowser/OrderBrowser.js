import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Typography } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Pagination from '@kevinwang0316/react-materialui-pagination';
// import Pagination from '../Pagination';

import OrderCard from './OrderCard';
import { MAX_ORDER_AMOUNT } from '../../config';

const styles = {
  root: {
    width: '90%',
    margin: '40px auto'
  },
  title: {
    fontSize: 16,
    fontWeight: 'bold'
  },
  cardDiv: {
    display: 'flex',
    flexWrap: 'wrap'
  }
};

export const OrderBrowser = ({
  orders, classes, title, menuItems, offset, totalOrder, onPageNumberClick
}) => (
  <div className={classes.root}>
    <Typography color="primary" className={classes.title}>{title}</Typography>
    {Object.keys(menuItems).length !== 0 && (
      <div className={classes.cardDiv}>
        {orders.map(order => <OrderCard order={order} menuItems={menuItems} key={order._id} />)}
      </div>
    )}
    <Pagination
      offset={offset}
      limit={MAX_ORDER_AMOUNT}
      total={totalOrder}
      onClick={onPageNumberClick}
    />
  </div>
);
OrderBrowser.prototype = {
  orders: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  menuItems: PropTypes.object.isRequired,
  onPageNumberClick: PropTypes.func.isRequired,
  offset: PropTypes.number.isRequired,
  totalOrder: PropTypes.number.isRequired
};
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems
});
export default connect(mapStateToProps, null)(withStyles(styles)(OrderBrowser));
