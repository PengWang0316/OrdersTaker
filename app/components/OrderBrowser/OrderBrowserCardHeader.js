import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { CardHeader, Tooltip } from '@material-ui/core';
import { lightGreen, yellow, grey, red } from '@material-ui/core/colors';
import { CheckCircle, OfflinePin, WatchLater, Help } from '@material-ui/icons';

import { ORDER_STATUS_FINISHED, ORDER_STATUS_RECEIVED, ORDER_STATUS_PREPARING } from '../../config';

const avatarIcon = {
  width: 40,
  height: 40
};

const styles = {
  checkCircle: {
    ...avatarIcon,
    color: lightGreen[700]
  },
  offlinPin: {
    ...avatarIcon,
    color: yellow[800]
  },
  watchLater: {
    ...avatarIcon,
    color: grey[500]
  },
  unknow: {
    ...avatarIcon,
    color: red[500]
  }
};

export const OrderBrowserCardHeader = ({ order, classes, ...rest }) => {
  const orderDate = new Date(order.dateStamp);
  let avatarElement;
  switch (order.status) {
    case ORDER_STATUS_FINISHED:
      avatarElement = (
        <Tooltip title="Finished" placement="right-end">
          <CheckCircle className={classes.checkCircle} />
        </Tooltip>
      );
      break;
    case ORDER_STATUS_RECEIVED:
      avatarElement = (
        <Tooltip title="Received" placement="right-end">
          <OfflinePin className={classes.offlinPin} />
        </Tooltip>
      );
      break;
    case ORDER_STATUS_PREPARING:
      avatarElement = (
        <Tooltip title="Preparing" placement="right-end">
          <WatchLater className={classes.watchLater} />
        </Tooltip>
      );
      break;
    default:
      avatarElement = (
        <Tooltip title="Unknow Status" placement="right-end">
          <Help className={classes.unknow} />
        </Tooltip>
      );
      break;
  }
  return (
    <CardHeader
      avatar={avatarElement}
      title={orderDate.toDateString()}
      subheader={orderDate.toLocaleTimeString()}
      {...rest}
    />
  );
};
OrderBrowserCardHeader.propTypes = {
  order: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
};
export default withStyles(styles)(OrderBrowserCardHeader);
