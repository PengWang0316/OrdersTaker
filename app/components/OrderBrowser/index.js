import OrderBrowser from './OrderBrowser';

export default OrderBrowser;
