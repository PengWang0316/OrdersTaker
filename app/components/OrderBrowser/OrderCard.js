import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Card, CardMedia, CardContent, Typography, IconButton, Tooltip } from '@material-ui/core';
import { DeleteForever, Link } from '@material-ui/icons';
import { red } from '@material-ui/core/colors';

import OrderBrowserCardHeader from './OrderBrowserCardHeader';
import OrdersPageContext from '../../contexts/OrdersPageContext';

const styles = {
  card: {
    minWidth: 300,
    maxWidth: 500,
    margin: '10px 10px',
    '&:hover': {
      cursor: 'pointer'
    }
  },
  media: {
    height: 0,
    paddingTop: '56.25%', // 16:9
  },
  cardContent: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-between'
  },
  iconButton: {
    width: 30,
    height: 30
  },
  deleteButton: {
    color: red[400]
  },
  icon: {
    width: 20,
    height: 20
  }
};

export const OrderCard = ({ order, classes, menuItems }) => {
  const itemsKey = Object.keys(order.items);
  return (
    <OrdersPageContext.Consumer>
      {({ handleShowDetail, handleRmoveUnloginOrder, handleLinkOrder }) => (
        <Card className={classes.card}>
          <OrderBrowserCardHeader order={order} onClick={() => handleShowDetail(order)} />
          <CardMedia className={classes.media} image={menuItems[itemsKey[0]].photo} title={menuItems[itemsKey[0]].name} onClick={() => handleShowDetail(order)} />
          <CardContent className={classes.cardContent}>
            <div>
              <Typography className={classes.title} color="textSecondary">This order has {itemsKey.length} items</Typography>
              <Typography color="textSecondary">Total Price: ${order.totalPrice}</Typography>
            </div>
            {!order.userId && (
              <div>
                <Tooltip title="Remove this order from this device">
                  <IconButton onClick={() => handleRmoveUnloginOrder(order._id)} className={`${classes.iconButton} ${classes.deleteButton}`}><DeleteForever className={classes.icon} /></IconButton>
                </Tooltip>
                <Tooltip title="Link this order to your account">
                  <IconButton onClick={() => handleLinkOrder(order._id)} color="primary" className={classes.iconButton}><Link className={classes.icon} /></IconButton>
                </Tooltip>
              </div>
            )}
          </CardContent>
        </Card>
      )}
    </OrdersPageContext.Consumer>
  );
};
OrderCard.propTypes = {
  order: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  menuItems: PropTypes.object.isRequired
};
export default withStyles(styles)(OrderCard);
