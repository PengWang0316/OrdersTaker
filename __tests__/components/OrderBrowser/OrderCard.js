import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { OrderCard } from '../../../app/components/OrderBrowser/OrderCard';
import context from '../../../app/contexts/OrdersPageContextTestHelper';


jest.mock('@material-ui/core/Card', () => 'Card');
jest.mock('@material-ui/core/CardMedia', () => 'CardMedia');
jest.mock('@material-ui/core/CardContent', () => 'CardContent');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/Tooltip', () => 'Tooltip');
jest.mock('@material-ui/icons/DeleteForever', () => 'DeleteForever');
jest.mock('@material-ui/icons/Link', () => 'Link');
jest.mock('../../../app/components/OrderBrowser/OrderBrowserCardHeader', () => 'OrderBrowserCardHeader');
jest.mock('../../../app/contexts/OrdersPageContext');

describe('OrderCard', () => {
  const defaultProps = {
    classes: {
      card: 'card',
      media: 'media',
      cardContent: 'cardContent',
      iconButton: 'iconButton',
      deleteButton: 'deleteButton',
      icon: 'icon'
    },
    order: {
      _id: 'ordersId',
      items: {
        orderIdA: {},
        orderIdB: {}
      },
      userId: 'userId'
    },
    menuItems: {
      orderIdA: { photo: 'orderA photo', name: 'orderA name' }
    },
    title: 'titel'
  };
  const getShallowComponent = (props = defaultProps) => mount(<OrderCard {...props} />);

  test('Snapshot with user id', () => expect(renderer.create(<OrderCard {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot without user id', () => expect(renderer.create(<OrderCard {...{
    ...defaultProps,
    order: {
      _id: 'ordersId',
      items: {
        orderIdA: {},
        orderIdB: {}
      },
      userId: null
    }
  }}
  />).toJSON()).toMatchSnapshot());

  test('Function called when click elements', () => {
    console.error = jest.fn();
    const newProps = {
      ...defaultProps,
      order: {
        _id: 'ordersId',
        items: {
          orderIdA: {},
          orderIdB: {}
        },
        userId: null
      }
    };
    const component = getShallowComponent(newProps);
    component.find('OrderBrowserCardHeader').at(0).simulate('click');
    expect(context.handleShowDetail).toHaveBeenCalledTimes(1);
    expect(context.handleShowDetail).toHaveBeenLastCalledWith(newProps.order);
    component.find('CardMedia').at(0).simulate('click');
    expect(context.handleShowDetail).toHaveBeenCalledTimes(2);
    expect(context.handleShowDetail).toHaveBeenLastCalledWith(newProps.order);
    component.find('IconButton').at(0).simulate('click');
    expect(context.handleRmoveUnloginOrder).toHaveBeenCalledTimes(1);
    expect(context.handleRmoveUnloginOrder).toHaveBeenLastCalledWith(newProps.order._id);
    component.find('IconButton').at(1).simulate('click');
    expect(context.handleLinkOrder).toHaveBeenCalledTimes(1);
    expect(context.handleLinkOrder).toHaveBeenLastCalledWith(newProps.order._id);
  });
});
