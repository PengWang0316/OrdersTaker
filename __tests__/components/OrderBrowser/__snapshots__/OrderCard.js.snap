// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderCard Snapshot with user id 1`] = `
<Card
  className="card"
>
  <OrderBrowserCardHeader
    onClick={[Function]}
    order={
      Object {
        "_id": "ordersId",
        "items": Object {
          "orderIdA": Object {},
          "orderIdB": Object {},
        },
        "userId": "userId",
      }
    }
  />
  <CardMedia
    className="media"
    image="orderA photo"
    onClick={[Function]}
    title="orderA name"
  />
  <CardContent
    className="cardContent"
  >
    <div>
      <Typography
        color="textSecondary"
      >
        This order has 
        2
         items
      </Typography>
      <Typography
        color="textSecondary"
      >
        Total Price: $
      </Typography>
    </div>
  </CardContent>
</Card>
`;

exports[`OrderCard Snapshot without user id 1`] = `
<Card
  className="card"
>
  <OrderBrowserCardHeader
    onClick={[Function]}
    order={
      Object {
        "_id": "ordersId",
        "items": Object {
          "orderIdA": Object {},
          "orderIdB": Object {},
        },
        "userId": null,
      }
    }
  />
  <CardMedia
    className="media"
    image="orderA photo"
    onClick={[Function]}
    title="orderA name"
  />
  <CardContent
    className="cardContent"
  >
    <div>
      <Typography
        color="textSecondary"
      >
        This order has 
        2
         items
      </Typography>
      <Typography
        color="textSecondary"
      >
        Total Price: $
      </Typography>
    </div>
    <div>
      <Tooltip
        title="Remove this order from this device"
      >
        <IconButton
          className="iconButton deleteButton"
          onClick={[Function]}
        >
          <DeleteForever
            className="icon"
          />
        </IconButton>
      </Tooltip>
      <Tooltip
        title="Link this order to your account"
      >
        <IconButton
          className="iconButton"
          color="primary"
          onClick={[Function]}
        >
          <Link
            className="icon"
          />
        </IconButton>
      </Tooltip>
    </div>
  </CardContent>
</Card>
`;
